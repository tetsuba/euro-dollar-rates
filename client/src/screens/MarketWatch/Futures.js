import React from 'react'
import axios from 'axios'
import SearchBox from '../../components/SearchBox/SearchBox'
import FilterList from '../../components/List/FilterList'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'

const URLS = {
  MARKET_WATCH: 'https://www.marketwatch.com/investing/future/',
}

export default class Futures extends React.Component {
  state = {
    list: [],
    text: ''
  }

  async componentDidMount() {
    const futures = await axios.get('http://localhost:3000/api/futures/list')
    const list = futures.data.list.map(({name}) => name)
    this.setState({
      list,
      group: this.groupList(list)
    })
  }

  groupList(list) {
    const j = list
      .filter((string) => !string.endsWith('00'))
      .map((string) => string.replace(string.slice(-3), ''))

    return [...new Set(j)]
  }

  inputText = (ele) => {
    this.setState({
      text: ele.target.value.toUpperCase()
    })
  }

  render() {
    return (
      <>
        <BreadCrumbs />
        <div className="row">
          <div className="col">
            <SearchBox
              text={this.state.text}
              label="Futures Symbol"
              inputFunc={this.inputText}
            />
          </div>
        </div>
        <div className="row mt-4">
          <div className="col">
            <FilterList
              label="Full List"
              list={this.state.list}
              text={this.state.text}
              link={URLS.MARKET_WATCH}
            />
          </div>
          <div className="col">
            <FilterList
              label="Group List"
              list={this.state.group}
              text={this.state.text}
            />
          </div>
        </div>
      </>
    )
  }
}
