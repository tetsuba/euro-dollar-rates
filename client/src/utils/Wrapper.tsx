import React from "react";
import {QueryClient, QueryClientProvider} from "react-query";
import {MemoryRouter} from "react-router-dom";

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry:0,
        }
    }
})

export default function Wrapper({pathname, children}: any) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={['', { pathname }]}>
                {children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}