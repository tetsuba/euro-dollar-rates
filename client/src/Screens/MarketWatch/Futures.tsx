import React, {useState} from 'react'
import {fetchFutures} from '../../utils/service'
import SearchBox from '../../Components/SearchBox/SearchBox'
import BreadCrumbs from '../../Components/BreadCrumbs/BreadCrumbs'
import FuturesFullList from '../../Components/List/FuturesFullList'
import FuturesGroupList from '../../Components/List/FuturesGroupList'
import { InputElementType } from '../../utils/types'
import {useQuery} from "react-query";
import LoadingSpinner from "../../Components/LoadingSpinner/LoadingSpinner"
import {filterFuturesList} from "../../utils/mutations"
import Error from "../Error"

const Futures = () => {
    const futures = useQuery(['futures'], fetchFutures)
    const [text, setText] = useState('')

    if (futures.isLoading) return <LoadingSpinner />
    if (futures.isError) return <Error />

    // TODO: is this the right way?
    const list = futures.data?.list || []
    const group = futures.data?.group || []

    return (
        <>
            <BreadCrumbs />
            <div className="row">
                <div className="col">
                    <SearchBox
                        text={text}
                        label="Futures Symbol"
                        inputFunc={(e: InputElementType) => setText(e.target.value.toUpperCase())}
                    />
                </div>
            </div>
            <div className="row mt-4">
                <div className="col">
                    <FuturesFullList
                        list={filterFuturesList(list, text)}
                        text={text}
                    />
                </div>
                <div className="col">
                    <FuturesGroupList
                        list={filterFuturesList(group, text)}
                        text={text}
                    />
                </div>
            </div>
        </>
    )
}

export default Futures
