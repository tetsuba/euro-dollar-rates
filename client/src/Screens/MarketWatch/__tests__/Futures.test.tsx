import React from 'react'
import {
    render,
    screen,
    waitForElementToBeRemoved,
    fireEvent, waitFor,
} from '@testing-library/react'

import Futures from '../Futures'
import Wrapper from "../../../utils/Wrapper";
import axios from 'axios'

jest.mock('axios')

const mockData = {
    date: '12/10/2022',
    list: [
        { name: 'EDT290', link: 'http://mock-lick.com' },
        { name: 'EDT240', link: 'http://mock-lick.com' },
        { name: 'EDT303', link: 'http://mock-lick.com' },
    ],
}



describe('Futures', () => {
    test('should render a loading screen until api call is completed', async () => {
        // @ts-ignore
        axios.get.mockResolvedValue({ data: mockData, status: 200 })
        const rendered = render(<Wrapper pathname="/market-watch/futures"><Futures /></Wrapper>)
        expect(rendered.asFragment()).toMatchSnapshot()
        await waitForElementToBeRemoved(() => screen.getByText(/Loading.../i))
        expect(rendered.asFragment()).toMatchSnapshot()
    })
    test('should render an error screen', async () => {
        // @ts-ignore
        axios.get.mockRejectedValueOnce()

        render(
            <Wrapper pathname="/market-watch/futures">
                <Futures />
            </Wrapper>
        )
        await waitFor(() => expect(screen.getByTestId('error-screen')))
    })

    describe('Event onChange search futures', () => {
        test('should render a list and a group list of futures', async () => {
            // @ts-ignore
            axios.get.mockResolvedValue({ data: mockData, status: 200 })
            render(<Wrapper pathname="/market-watch/futures"><Futures /></Wrapper>)

            await waitForElementToBeRemoved(() => screen.getByText(/Loading.../i))
            fireEvent.change(screen.getByTestId('search-input'), {
                target: { value: 'E' },
            })

            expect(screen.queryAllByTestId('external-link')).toHaveLength(3)
            expect(screen.queryAllByTestId('internal-link')).toHaveLength(1)
        })
    })
})
