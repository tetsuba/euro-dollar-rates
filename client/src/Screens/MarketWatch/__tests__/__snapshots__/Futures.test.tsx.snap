// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Futures should render a loading screen until api call is completed 1`] = `
<DocumentFragment>
  <div
    class="spinner-border"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;

exports[`Futures should render a loading screen until api call is completed 2`] = `
<DocumentFragment>
  <div
    class="row shadow mb-5"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb mt-3"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumb-item"
        >
          <a
            href="/market-watch"
          >
            market-watch
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          futures
        </li>
      </ol>
    </nav>
  </div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div
        class="input-group input-group-lg"
      >
        <span
          class="input-group-text"
          id="inputGroup-sizing-lg"
        >
          Futures Symbol
        </span>
        <input
          aria-describedby="inputGroup-sizing-lg"
          aria-label="Sizing example input"
          class="form-control"
          data-testid="search-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class="row mt-4"
  >
    <div
      class="col"
    />
    <div
      class="col"
    />
  </div>
</DocumentFragment>
`;
